includes/class-events-store.php

```diff
@@ -0,0 +1,592 @@
+
+class Events_Store extends Singleton {
+	/**
+	 * PLUGIN SETUP
+	 */
+
+	/**
+	 * Class properties
+	 */
+	const TABLE_SUFFIX = 'a8c_cron_control_jobs';
+
+	const DB_VERSION        = 1;
+	const DB_VERSION_OPTION = 'a8c_cron_control_db_version';
+	const TABLE_CREATE_LOCK = 'a8c_cron_control_creating_table';
+
+	const STATUS_PENDING   = 'pending';
+	const STATUS_RUNNING   = 'running';
+	const STATUS_COMPLETED = 'complete';
+
+	const CACHE_KEY = 'a8c_cron_ctrl_option';
+
+	private $job_creation_suspended = false;
+
+	/**
+	 * Register hooks
+	 */
+	protected function class_init() {
+		// Check that the table exists and is the correct version
+		$this->prepare_tables();
+
+		// Option interception
+		add_filter( 'pre_option_cron', array( $this, 'get_option' ) );
+		add_filter( 'pre_update_option_cron', array( $this, 'update_option' ), 10, 2 );
+
+		// Disallow duplicates
+		add_filter( 'schedule_event', array( $this, 'block_creation_if_job_exists' ) );
+	}
+
+	/**
+	 * Build appropriate table name for this install
+	 */
+	public function get_table_name() {
+		global $wpdb;
+
+		return $wpdb->prefix . self::TABLE_SUFFIX;
+	}
+
+	/**
+	 * Build array of valid event statuses
+	 */
+	public function get_allowed_statuses() {
+		return array(
+			self::STATUS_PENDING,
+			self::STATUS_RUNNING,
+			self::STATUS_COMPLETED,
+		);
+	}
+
+	/**
+	 * Create the plugin's DB table when necessary
+	 */
+	protected function prepare_tables() {
+		// Nothing to do
+		if ( (int) get_option( self::DB_VERSION_OPTION ) === self::DB_VERSION ) {
+			return;
+		}
+
+		// Limit chance of race conditions when creating table
+		if ( false === wp_cache_add( self::TABLE_CREATE_LOCK, 1, null, 1 * \MINUTE_IN_SECONDS ) ) {
+			return;
+		}
+
+		// Use Core's method of creating/updating tables
+		if ( ! function_exists( 'dbDelta' ) ) {
+			require_once ABSPATH . '/wp-admin/includes/upgrade.php';
+		}
+
+		global $wpdb;
+
+		// Define schema and create the table
+		$schema = "CREATE TABLE `{$this->get_table_name()}` (
+			`ID` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
+
+			`timestamp` bigint(20) unsigned NOT NULL,
+			`action` varchar(255) NOT NULL,
+			`action_hashed` varchar(32) NOT NULL,
+			`instance` varchar(32) NOT NULL,
+
+			`args` longtext NOT NULL,
+			`schedule` varchar(255) DEFAULT NULL,
+			`interval` int unsigned DEFAULT 0,
+			`status` varchar(32) NOT NULL DEFAULT 'pending',
+
+			`created` datetime NOT NULL,
+			`last_modified` datetime NOT NULL,
+
+			PRIMARY KEY (`ID`),
+			UNIQUE KEY `ts_action_instance` (`timestamp`, `action` (191), `instance`, `status`)
+		) ENGINE=InnoDB;\n";
+
+		dbDelta( $schema, true );
+
+		// Confirm that the table was created, and set the option to prevent further updates
+		$table_count = count( $wpdb->get_col( "SHOW TABLES LIKE '{$this->get_table_name()}'" ) );
+
+		if ( 1 === $table_count ) {
+			update_option( self::DB_VERSION_OPTION, self::DB_VERSION, true );
+		} else {
+			delete_option( self::DB_VERSION_OPTION );
```

Comment:
Yeah, I've been unsure of what to do here, because if the table doesn't exist, cron is completely broken as long as the plugin is active. If the table isn't created, the plugin shouldn't override `get|update_option()` calls. Setting an option or cache key to disable the filters in `class_init()` might be worthwhile, as well as setting the option to the expected version as a short-circuit to future upgrades. An action, which we can hook separately from mu-plugins, could handle alerting.